apps/native-component-list/src/screens/AV/AudioPlayer.tsx

```diff
@@ -128,7 +152,74 @@ export default class AudioPlayer extends React.Component<Props, State> {
         setRateAsync={this._setRateAsync}
         setIsMutedAsync={this._setIsMutedAsync}
         setVolume={this._setVolumeAsync}
+        extraIndicator={<JsiAudioBar isPlaying={this.state.isPlaying} sound={this._sound!} />}
       />
     );
   }
 }
+
+// for some reason, iOS returns much smaller sample values
+const inputRange = Platform.OS === 'ios' ? [0, 0.3] : [0, 1];
+
+function JsiAudioBar({ sound, isPlaying }: { sound: Audio.Sound; isPlaying: boolean }) {
+  const sharedValue = useSharedValue(0);
+  const animatedStyle = useAnimatedStyle(() => {
+    const barWidth = interpolate(sharedValue.value, inputRange, [1, 500], Extrapolate.CLAMP);
+    return {
+      width: withSpring(barWidth, {
+        mass: 1,
+        damping: 500,
+        stiffness: 1000,
+      }),
+    };
+  }, [sharedValue]);
+
+  const isJsiAudioSupported = React.useMemo(() => {
+    try {
+      // @ts-expect-error that method is private
+      sound?._updateAudioSampleReceivedCallback();
+      return true;
+    } catch (e: unknown) {
+      if (e instanceof UnavailabilityError) {
+        return false;
+      }
+      throw e;
+    }
+  }, [sound]);
+
+  React.useEffect(() => {
+    if (isJsiAudioSupported && isPlaying && !sound?._onAudioSampleReceived) {
+      sound?.setOnAudioSampleReceived((sample) => {
+        const frames = sample.channels[0].frames;
+        const frameSum = frames.slice(0, 200).reduce((prev, curr) => prev + curr ** 2, 0);
+        const rmsValue = Math.sqrt(frameSum / 200);
+
+        runOnUI(() => {
+          sharedValue.value = rmsValue;
+        })();
```

Comment:
Correct me if I'm wrong here, but the whole method body does the following:
1. upon each audio sample invoke callback on JS thread (via bridge)
2. calculate what's needed to be calculated
3. pass the calculation result to the reanimated thread by invoking `runOnUI` method (via bridge)

I guess the optimal solution would be to perform all the operations on reanimated thread; something like this:
```ts
sound?.setOnAudioSampleReceived((sample) => {
  'worklet'
  const frames = sample.channels[0].frames;
  const frameSum = frames.slice(0, 200).reduce((prev, curr) => prev + curr ** 2, 0);
  const rmsValue = Math.sqrt(frameSum / 200);
  sharedValue.value = rmsValue;
});
```
Dunno if that's doable 🤷

Reply:
From my experience of creating my demo app - setting shared values so often in a thread different than UI thread sometimes lead to weird race conditions and crash the app randomly.

Anyway, according to [reanimated docs](https://docs.swmansion.com/react-native-reanimated/docs/2.2.0/shared-values/), the shared value is gonna be set asynchronously in either case:
> By updating .value property. Updating .value is sync when running on the UI thread, or async when running on the React Native JS thread.

Reply:
I was just wondering about this cycle that is happening here:
1. call the js callback from the native code
2. pass the data via bridge (I wonder if that sentence is true though 😅 Is crossing the bridge happening at all here? Because you're working on using JSI here right, so no bridge involved?)
3. do some work on JS thread (just simple calculations)
4. update the shared value in async manner (I wonder how much asynchronous is this? Is it just a one-frame-duration gap or it's rather hard to determine this? What happens when the bridge is clogged up with the messages?)

There's so much I don't understand here and I guess I should not worry this much anyway 😅

Reply:
You are mostly right, except that there's no bridge here 😁

1. Some native audio driver receives data and passes them to our JS callback via JSI. It's not totally synchronous (it couldn't be, because the audio data is received on some different thread than JS is running on), but also not via the bridge. The key here is the "js call invoker" - an object which schedules functions from any c++ thread to be run on the js thread. 
2. the JS callback is invoked from the c++ code, does some calculations
3. (I'm not reanimated expert) it schedules the job to be run on UI thread, mostly it happens the similar way, e.g. for ios somewhere there's called `dispatch_async(dispatch_get_main_queue(), ^{ ... })`
4.  [see this code](https://github.com/software-mansion/react-native-reanimated/blob/008ee9322f3d2ed3ffef1996cbab50039fb72a3d/Common/cpp/SharedItems/MutableValue.cpp?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L43). When looking at this code, you can see that it does exactly what the citated docs above say. _And the only reason I used `runOnUI()` is that I found it more stable than doing it from different thread. Maybe that's a reanimated bug? I'll try to dig deeper when I have more time, but for now I'm gonna keep it as is._ 😉

Reply:
Thank you very much for detailed explanation 🤩 🚀